src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs

```diff
@@ -14,290 +14,435 @@ namespace Microsoft.CodeAnalysis.CSharp
+                // Start with the part of the decision tree that is in scope of the section variables.
+                // Its endpoint is not reachable (it jumps back into the decision tree code).
+                var sectionBuilder = pslr.SwitchSections[section];
+
+                // Add labels corresponding to the labels of the switch section.
                 foreach (var label in section.SwitchLabels)
                 {
-                    if (usedLabels.Contains(label.Label))
-                    {
-                        if (sectionBuilder == null)
-                        {
-                            sectionBuilder = ArrayBuilder<BoundStatement>.GetInstance();
-                        }
-
-                        sectionBuilder.Add(_factory.Label(label.Label));
-                    }
+                    sectionBuilder.Add(_factory.Label(label.Label));
                 }
 
-                if (sectionBuilder != null)
-                {
-                    sectionBuilder.AddRange(VisitList(section.Statements));
-                    sectionBuilder.Add(_factory.Goto(node.BreakLabel));
-                    result.Add(_factory.Block(section.Locals, sectionBuilder.ToImmutableAndFree()));
-                }
+                // Add the translated body of the switch section
+                sectionBuilder.AddRange(VisitList(section.Statements));
+                sectionBuilder.Add(_factory.Goto(node.BreakLabel));
+                result.Add(_factory.Block(section.Locals, sectionBuilder.ToImmutableAndFree()));
+                // at this point the end of result is unreachable.
             }
 
             result.Add(_factory.Label(node.BreakLabel));
-            return _factory.Block(usedTemps.ToImmutableArray().Concat(node.InnerLocals), node.InnerLocalFunctions, result.ToImmutableAndFree());
+            var translatedSwitch = _factory.Block(pslr.DeclaredTemps.ToImmutableArray().Concat(node.InnerLocals), node.InnerLocalFunctions, result.ToImmutableAndFree());
+            return translatedSwitch;
         }
 
-        private void LowerDecisionTree(DecisionTree decisionTree, HashSet<LabelSymbol> usedLabels, HashSet<LocalSymbol> usedTemps, ArrayBuilder<BoundStatement> result)
+        private class PatternSwitchLocalRewriter
         {
-            if (decisionTree == null) return;
-            switch (decisionTree.Kind)
+            public readonly LocalRewriter LocalRewriter;
+            public readonly HashSet<LocalSymbol> DeclaredTempSet = new HashSet<LocalSymbol>();
+            public readonly ArrayBuilder<LocalSymbol> DeclaredTemps = ArrayBuilder<LocalSymbol>.GetInstance();
+            public readonly Dictionary<BoundPatternSwitchSection, ArrayBuilder<BoundStatement>> SwitchSections = new Dictionary<BoundPatternSwitchSection, ArrayBuilder<BoundStatement>>();
+            public ArrayBuilder<BoundStatement> LoweredDecisionTree = ArrayBuilder<BoundStatement>.GetInstance();
```

Comment:
I've added the missing "Free".

It is contained in a private class. It is only "public" to the containing type.

---
In reply to: [70303266](https://github.com/dotnet/roslyn/pull/12416#discussion_r70303266) [](ancestors = 70303266)